Codegen: Generate static dispatch headers per operator

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #68462
* #68693
* #68692
* #68691
* #68697
* #68690
* #68689
* __->__ #68714
* #68688
* #68687
* #68247

This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

peterbell10 committed:
Codegen: Generate static dispatch headers per operator

This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

[ghstack-poisoned]

albanD commented on aten/src/ATen/templates/DispatchKeyFunctions_inl.h:
```diff
@@ -11,14 +12,8 @@
-
-} // namespace ${dispatch_namespace}
-} // namespace at
+$DispatchKeyFunctions_inl_includes
```
Missing {} ?

peterbell10 replied:
They're actually optional:
https://github.com/pytorch/pytorch/blob/9554ebe44e6e73dc75105d4935d41e626e03299b/tools/codegen/code_template.py#L4

albanD replied:
Right, but that's the only place where we don't use them? :p 
Not very important.

albanD reviewed (commented):
Sound good, minor typo.

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).

peterbell10 committed:
Rebase to fix merge conflicts on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Fix mypy issues on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Commit to re-run CI on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Generate static dispatch headers per operator"


This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).
